ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,882 @@
+/* -*- Mode: C; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (C) 2024, Advanced Micro Devices, Inc. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "mpi.h"
+#include "ompi/communicator/communicator.h"
+#include "ompi/mca/coll/base/coll_base_functions.h"
+#include "opal/include/opal/align.h"
+
+static inline void *coll_acoll_malloc(coll_acoll_reserve_mem_t *reserve_mem_ptr, uint64_t size)
+{
+    void *temp_ptr = NULL;
+    if ((true == reserve_mem_ptr->reserve_mem_allocate)
+        && (size <= reserve_mem_ptr->reserve_mem_size)
+        && (false == reserve_mem_ptr->reserve_mem_in_use)) {
+        if (NULL == reserve_mem_ptr->reserve_mem) {
+            reserve_mem_ptr->reserve_mem = malloc(reserve_mem_ptr->reserve_mem_size);
+        }
+        temp_ptr = reserve_mem_ptr->reserve_mem;
+
+        if (NULL != temp_ptr) {
+            reserve_mem_ptr->reserve_mem_in_use = true;
+        }
+    } else {
+        temp_ptr = malloc(size);
+    }
+
+    return temp_ptr;
+}
+
+static inline void coll_acoll_free(coll_acoll_reserve_mem_t *reserve_mem_ptr, void *ptr)
+{
+    if ((false == reserve_mem_ptr->reserve_mem_allocate)
+        || (false == reserve_mem_ptr->reserve_mem_in_use)) {
+        if (NULL != ptr) {
+            free(ptr);
```

Comment:
Please see the reply to the previous comment on coll_acoll_reserve_mem_t. Since ptr will be the same as reserve_mem (the pre-allocated memory) when the memory size needed is less than or equal to that of the pre-allocated memory, the null check should be inside to ensure it is the case corresponding to a new buffer.